module/mempool/stdmap/eject.go

```diff
@@ -165,7 +165,7 @@ func (q *LRUEjector) Eject(b *Backend) (flow.Identifier, flow.Entity, bool) {
 	// finds the oldest entity
 	oldestSQ := uint64(math.MaxUint64)
 	var oldestID flow.Identifier
-	for id := range b.backData.All() {
```

Comment:
Am a bit confused about this change. In `eject.go` then there are two other functions -- `EjectTrueRandom()` and `EjectTrueRandomFast()` -- which also do `range b.backData.All()` but those functions have not been refactored, only this one?

Also, presumably none of `Eject()`, `EjectTrueRandom()`, or `EjectTrueRandomFast()` are used anyway with the new `heroCache` efficient eject mechanisms, or? So why refactor these at all?

Also, how to to ensure that they are not accidentally used in addition to the new `heroCache` efficient eject mechanisms?

Reply:
> Also, presumably none of Eject(), EjectTrueRandom(), or EjectTrueRandomFast() are used anyway with the new heroCache efficient eject mechanisms, or? So why refactor these at all?

Depending on their functionality some ejections need only the entities, some only the identifiers, and some both. When both entities and identifiers are needed, the `All` is utilized, and in other cases only the relevant methods. Note that the entities may be stored with a different identifier than their `entity.ID()`. Hence, for those applications that require both identifiers and entities, we cannot derive the former from the latter. 

Correct, `eject` methods are idempotent on `heroCache`, however, since `MapBackData` is utilizing them, it can benefit from this refactor, i.e., to have a slice instead of the map as much as places as possible. 

> Also, how to to ensure that they are not accidentally used in addition to the new heroCache efficient eject mechanisms?

The `Backend` which is the controller layer of `mempools` and runs on top of a `backdata` only kicks in the `eject` [if after adding an entity to the backdata, its _size_ goes beyond its _capacity_](https://github.com/onflow/flow-go/blob/c9eb546e78bf3f62af8b00798d9d7ed11b460bc5/module/mempool/stdmap/backend.go#L209). This is only applicable to the `MapBackData` that lets the size of map grows and then scrapes any excess entities if it goes beyond its `guaranteedCapacity` parameter. Please note that this will _never happen_ in `heroCache` (which is also an implementation of `backdata`. Why? because `heroCache` always conducts the size adjustment as part of the `Add` method. So, it guarantees that by the time its `Add` method returns, the size of `heroCache` is always less than or equal to its `sizeLimit`. Hence [this if statement](https://github.com/onflow/flow-go/blob/c9eb546e78bf3f62af8b00798d9d7ed11b460bc5/module/mempool/stdmap/backend.go#L209) is always false for `heroCache`.

Reply:
Ah, that makes sense. The old `Add()` method first makes the `backdata` larger and then prunes it when it gets too big. But `heroCache` never gets too big. Very good! :-)